File: go/lambda-api-ci/lib/cistack/cistack.go
package cistack

import (
	"fmt"

	"github.com/aws/aws-cdk-go/awscdk"
	"github.com/aws/aws-cdk-go/awscdk/awscodebuild"
	"github.com/aws/aws-cdk-go/awscdk/awscodecommit"
	"github.com/aws/aws-cdk-go/awscdk/awscodepipeline"
	"github.com/aws/aws-cdk-go/awscdk/awscodepipelineactions"
	"github.com/aws/aws-cdk-go/awscdk/awsiam"
	"github.com/aws/jsii-runtime-go"
)

type CIStack awscdk.Stack

type CIStackProps struct {
	awscdk.StackProps
	RepositoryName     *string
	LambdaApiStackName *string
	LambdaFunctionName *string
}

func NewCIStack(scope awscdk.Construct, id string, props CIStackProps) CIStack {
	stack := awscdk.NewStack(scope, &id, &props.StackProps)

	pipeline := awscodepipeline.NewPipeline(stack, jsii.String("Pipeline"), nil)

	repo := awscodecommit.Repository_FromRepositoryName(stack, jsii.String("WidgetsServiceRepository"), props.RepositoryName)
	sourceOutput := awscodepipeline.NewArtifact(jsii.String("SourceOutput"))
	sourceAction := awscodepipelineactions.NewCodeCommitSourceAction(&awscodepipelineactions.CodeCommitSourceActionProps{
		ActionName: jsii.String("CodeCommit"),
		Repository: repo,
		Output:     sourceOutput,
	})
	pipeline.AddStage(&awscodepipeline.StageOptions{
		StageName: jsii.String("Source"),
		Actions:   &[]awscodepipeline.IAction{sourceAction},
	})

	createBuildStage(stack, props, pipeline, sourceOutput)

	return stack
}

func createBuildStage(
	stack CIStack,
	props CIStackProps,
	pipeline awscodepipeline.Pipeline,
	sourceOutput awscodepipeline.Artifact,
) awscodepipeline.Artifact {
	project := awscodebuild.NewPipelineProject(stack, jsii.String("BuildProject"), &awscodebuild.PipelineProjectProps{
		Environment: &awscodebuild.BuildEnvironment{
			BuildImage: awscodebuild.LinuxBuildImage_STANDARD_3_0(),
		},
	})

	cdkDeployPolicy := awsiam.NewPolicyStatement(&awsiam.PolicyStatementProps{
		Actions: &[]*string{
			jsii.String("cloudformation:GetTemplate"),
			jsii.String("cloudformation:CreateChangeSet"),
			jsii.String("cloudformation:DescribeChangeSet"),
			jsii.String("cloudformation:ExecuteChangeSet"),
			jsii.String("cloudformation:DescribeStackEvents"),
			jsii.String("cloudformation:DeleteChangeSet"),
			jsii.String("cloudformation:DescribeStacks"),
			jsii.String("s3:*Object"),
			jsii.String("s3:ListBucket"),
			jsii.String("s3:getBucketLocation"),
			jsii.String("lambda:UpdateFunctionCode"),
			jsii.String("lambda:GetFunction"),
			jsii.String("lambda:CreateFunction"),
			jsii.String("lambda:DeleteFunction"),
			jsii.String("lambda:GetFunctionConfiguration"),
			jsii.String("lambda:AddPermission"),
			jsii.String("lambda:RemovePermission"),
		},
		Resources: &[]*string{
			stack.FormatArn(&awscdk.ArnComponents{
				Service:      jsii.String("cloudformation"),
				Resource:     jsii.String("stack"),
				ResourceName: jsii.String("CDKToolkit/*"),
			}),
			stack.FormatArn(&awscdk.ArnComponents{
				Service:      jsii.String("cloudformation"),
				Resource:     jsii.String("stack"),
				ResourceName: jsii.String(fmt.Sprintf("%s/*", *props.LambdaApiStackName)),
			}),
			stack.FormatArn(&awscdk.ArnComponents{
				Service:      jsii.String("lambda"),
				Resource:     jsii.String("function"),
				Sep:          jsii.String(":"),
				ResourceName: props.LambdaFunctionName,
			}),
			jsii.String("arn:aws:s3:::cdktoolkit-stagingbucket-*"),
		},
	})

	editOrCreateLambdaDependencies := awsiam.NewPolicyStatement(&awsiam.PolicyStatementProps{
		Actions: &[]*string{
			jsii.String("iam:GetRole"),
			jsii.String("iam:PassRole"),
			jsii.String("iam:CreateRole"),
			jsii.String("iam:AttachRolePolicy"),
			jsii.String("iam:PutRolePolicy"),
			jsii.String("apigateway:GET"),
			jsii.String("apigateway:DELETE"),
			jsii.String("apigateway:PUT"),
			jsii.String("apigateway:POST"),
			jsii.String("apigateway:PATCH"),
			jsii.String("s3:CreateBucket"),
			jsii.String("s3:PutBucketTagging"),
		},
		Resources: &[]*string{jsii.String("*")},
	})

	project.AddToRolePolicy(cdkDeployPolicy)
	project.AddToRolePolicy(editOrCreateLambdaDependencies)

	buildOutput := awscodepipeline.NewArtifact(jsii.String("BuildOutput"))
	buildAction := awscodepipelineactions.NewCodeBuildAction(&awscodepipelineactions.CodeBuildActionProps{
		ActionName: jsii.String("Build"),
		Project:    project,
		Input:      sourceOutput,
		Outputs:    &[]awscodepipeline.Artifact{buildOutput},
	})

	pipeline.AddStage(&awscodepipeline.StageOptions{
		StageName: jsii.String("build"),
		Actions:   &[]awscodepipeline.IAction{buildAction},
	})

	return buildOutput
}

File: typescript/lambda-api-ci/lib/ci-stack.ts
import { CodeCommitSourceAction, CodeBuildAction } from "aws-cdk-lib/aws-codepipeline-actions"
import { PolicyStatement } from "aws-cdk-lib/aws-iam"
import { ArnFormat, Stack, StackProps } from "aws-cdk-lib"
import { PipelineProject, LinuxBuildImage } from "aws-cdk-lib/aws-codebuild"
import { Artifact, Pipeline } from "aws-cdk-lib/aws-codepipeline"
import { Repository } from "aws-cdk-lib/aws-codecommit"
import { Construct } from "constructs"
import { lambdaApiStackName, lambdaFunctionName } from "../bin/lambda"

interface CIStackProps extends StackProps {
    repositoryName: string
}

export class CIStack extends Stack {
    constructor(scope: Construct, name: string, props: CIStackProps) {
        super(scope, name, props)

        const pipeline = new Pipeline(this, "Pipeline", {})

        const repo = Repository.fromRepositoryName(
            this,
            "WidgetsServiceRepository",
            props.repositoryName,
        )
        const sourceOutput = new Artifact("SourceOutput")
        const sourceAction = new CodeCommitSourceAction({
            actionName: "CodeCommit",
            repository: repo,
            output: sourceOutput,
            branch: "main",
        })
        pipeline.addStage({
            stageName: "Source",
            actions: [sourceAction],
        })

        this.createBuildStage(pipeline, sourceOutput)
    }

    private createBuildStage(pipeline: Pipeline, sourceOutput: Artifact) {
        const project = new PipelineProject(this, `BuildProject`, {
            environment: {
                buildImage: LinuxBuildImage.STANDARD_7_0,
            },
        })
        const cdkAssumeRolePolicy = new PolicyStatement()
        cdkAssumeRolePolicy.addActions("sts:AssumeRole")
        cdkAssumeRolePolicy.addResources(
            this.formatArn({
                service: "iam",
                resource: "role",
                region: "",
                resourceName: "cdk-*",
            }),
        )
        const cdkDeployPolicy = new PolicyStatement()
        cdkDeployPolicy.addActions(
            "cloudformation:GetTemplate",
            "cloudformation:CreateChangeSet",
            "cloudformation:DescribeChangeSet",
            "cloudformation:ExecuteChangeSet",
            "cloudformation:DescribeStackEvents",
            "cloudformation:DeleteChangeSet",
            "cloudformation:DescribeStacks",
            "s3:*Object",
            "s3:ListBucket",
            "s3:getBucketLocation",
            "lambda:UpdateFunctionCode",
            "lambda:GetFunction",
            "lambda:CreateFunction",
            "lambda:DeleteFunction",
            "lambda:GetFunctionConfiguration",
            "lambda:AddPermission",
            "lambda:RemovePermission",
            "ssm:GetParameter",
        )
        cdkDeployPolicy.addResources(
            this.formatArn({
                service: "cloudformation",
                resource: "stack",
                resourceName: "CDKToolkit/*",
            }),
            this.formatArn({
                service: "cloudformation",
                resource: "stack",
                resourceName: `${lambdaApiStackName}/*`,
            }),
            this.formatArn({
                service: "lambda",
                resource: "function",
                arnFormat: ArnFormat.COLON_RESOURCE_NAME,
                resourceName: lambdaFunctionName,
            }),
            this.formatArn({
                service: "ssm",
                resource: "parameter",
                resourceName: "cdk-bootstrap/*",
            }),
            "arn:aws:s3:::cdktoolkit-stagingbucket-*",
        )
        const editOrCreateLambdaDependencies = new PolicyStatement()
        editOrCreateLambdaDependencies.addActions(
            "iam:GetRole",
            "iam:PassRole",
            "iam:CreateRole",
            "iam:AttachRolePolicy",
            "iam:PutRolePolicy",
            "apigateway:GET",
            "apigateway:DELETE",
            "apigateway:PUT",
            "apigateway:POST",
            "apigateway:PATCH",
            "s3:CreateBucket",
            "s3:PutBucketTagging",
        )
        editOrCreateLambdaDependencies.addResources("*")
        project.addToRolePolicy(cdkDeployPolicy)
        project.addToRolePolicy(editOrCreateLambdaDependencies)
        project.addToRolePolicy(cdkAssumeRolePolicy)

        const buildOutput = new Artifact(`BuildOutput`)
        const buildAction = new CodeBuildAction({
            actionName: `Build`,
            project,
            input: sourceOutput,
            outputs: [buildOutput],
        })

        pipeline.addStage({
            stageName: "build",
            actions: [buildAction],
        })

        return buildOutput
    }
}
